### DISCUSSION TITLE
Is it possible to set absolute mark's size in 3D scatter?

### DISCUSSION THREAD

**Discussion by triuk** [alandefreitas/matplotplusplus#254] [Q&A]
Hi, I've just noticed that circles representing points in 3D scatter change their absolute size according the "situation on the screen".
As I'd like to draw particles of exact size in the plot - **is it possible to somehow fix the mark's render size?**
I know I have a specific use case, so thanks for the help again.

**Comment by triuk** [comment]
OK since I could not figure out the relationship between mark size and axis size and figure size, I decided to render spherical surface of radius R for each XYZ datapoint. I know it'd be heavy load at 100+ points, but this is the best I came with :/
Apparently there are two types I can use: `fsurf` and `fmesh`, unfortunately I can't make a single color sphere out of either one. I've tried many parameters, but none of them got rid of the black lines at fsurf or filled the empty space at fmesh.
- I looked at the source code, searching for some hint, but after several hours I am completely out of ideas, how to make the sphere single color.
- The second problem is the sphere's color. I must miss the plotting order somehow, it always stays at the last colormap I define.

Could you please advice me on this? There is probably a small thing I am missing, it's always like this.
```cpp
void plotParticle (double x, double y, double z, double r) {   
    auto funx = [r,x](double u, double v) { return r * cos(u) * sin(v)+x; };
    auto funy = [r,y](double u, double v) { return r * sin(u) * sin(v)+y; };
    auto funz = [r,z](double u, double v) { return r * cos(v)+z; };
    auto g1 = matplot::fsurf(funx, funy, funz, {0, 2*matplot::pi, 0, matplot::pi}, "", 10);
    //auto g1 = matplot::fmesh(funx, funy, funz, std::array<double, 4>{0, 2 * matplot::pi, 0, matplot::pi});
}

int main() {
    double AxLimit = 1000;
    
    auto fig = matplot::figure(false);
    fig->size(1000,1000);
    auto graph = matplot::gca();
    matplot::xlim({0, AxLimit});  matplot::ylim({0, AxLimit});  matplot::zlim({0, AxLimit});
    matplot::box(matplot::on);    graph->box_full(true);

    matplot::hold(matplot::on);
    matplot::colormap({{0, 0, 1}});
    plotParticle(100,100,100,100);
    matplot::colormap({{0, 1, 0}});
    plotParticle(500,500,500,200);
    matplot::hold(matplot::off);

    matplot::show();
    return 0;
}
```

**Reply by alandefreitas** [reply]
By "situation on the screen" you mean the 3d perspective/projection? Any images?

**Comment by triuk** [comment]
First thing is marks are always round, not elliptical - no problem, just set fig resolution to the square and good to go. All graphs contain the same data and I projected it to XZ plane for better comparison. The mark size depends on axis limits, filetype and fig size.
These are results of save(); with show() I do not have even the slightest idea, what's going on, because it shows different image ratio than is actually saved :D

**size(1000,1000); Axis limit(0,1000)**
JPG
![Fig1000x1000_axis1000](https://user-images.githubusercontent.com/93537202/161391341-3bf78e0c-5f7f-4245-a80d-41f6125b1fdc.jpg)
**size(1000,1000); Axis limit(0,400)**
PNG
![Fig1000x1000_axis400](https://user-images.githubusercontent.com/93537202/161391371-8ac60cfd-81ac-42cb-af25-88372af7a517.png)
SVG
![Fig1000x1000_axis400](https://user-images.githubusercontent.com/93537202/161391381-3f5fbfdb-4cbe-41e1-a952-f048b54d35e3.svg)
GIF
![Fig1000x1000_axis400](https://user-images.githubusercontent.com/93537202/161391385-7b22417c-3950-4202-bbed-40a4a4a56029.gif)
JPG
![Fig1000x1000_axis400](https://user-images.githubusercontent.com/93537202/161391361-f2ebe5b9-6a68-4a45-b20e-ce25ab5245dc.jpg)
**size(500,500); Axis limit(0,400)**
JPG
![Fig500x500_axis400](https://user-images.githubusercontent.com/93537202/161391401-1dc1da8e-e77e-4638-b50f-649e9e8e35da.jpg)

**Comment by alandefreitas** [comment]
Yes. The point of circles and marks is to have something that is still circular, and not elliptical, regardless of the ratio, which is quite hard to achieve otherwise. For something elliptical, you have to draw them separately.

**Reply by triuk** [reply]
I get it, marks are marks. If I set fig dimensions and xyz axis to square, it could represent real spherical particle. I can understand this just fine.
The point is, they do not represent the same "real size" - once they are below 50 in radius, once above. This is what I meant by "situation on the screen". And I most likely will not be able to comprehend the render pattern, to calculate needed mark's size to represent real particle of radius R. OK, I know my use case is very specific and this is probably not the way how to do it.
So I decided to render 3D spherical surface instead and I have two problems I can not deal with right now - make the sphere single color and be able to render spheres at the color of my choice in the same fig.
I raised the issue at the first "answer" with code. Could you please look at it? I am really out of ideas right now :/
